daemon/containerd/image_import.go

```diff
@@ -1,18 +1,371 @@
+	if err != nil {
+		logger.WithError(err).Debug("failed to write layer blob")
+		return "", err
+	}
+	logger = logrus.WithField("compressedDigest", compressedDigest)
+	logger = logger.WithField("uncompressedDigest", uncompressedDigest)
+
+	size, err := fillUncompressedLabel(ctx, cs, compressedDigest, uncompressedDigest)
+	if err != nil {
+		logger.WithError(err).Debug("failed to write compressed rootfs blob")
+		return "", err
+	}
+
+	compressedRootfsDesc := ocispec.Descriptor{
+		MediaType: mt,
+		Digest:    compressedDigest,
+		Size:      size,
+	}
+
+	cfg, err := dockerfile.BuildFromConfig(ctx, &container.Config{}, changes, platform.OS)
+	if err != nil {
+		logger.WithError(err).Debug("failed to process changes")
+		return "", errdefs.InvalidParameter(err)
+	}
+
+	ociCfg := containerConfigToOciImageConfig(cfg)
+	createdAt := time.Now()
+	config := ocispec.Image{
+		Architecture: platform.Architecture,
+		OS:           platform.OS,
+		Created:      &createdAt,
+		Author:       "",
+		Config:       ociCfg,
+		RootFS: ocispec.RootFS{
+			Type:    "layers",
+			DiffIDs: []digest.Digest{uncompressedDigest},
+		},
+		History: []ocispec.History{
+			{
+				Created:    &createdAt,
+				CreatedBy:  "",
+				Author:     "",
+				Comment:    msg,
+				EmptyLayer: false,
+			},
+		},
+	}
+	configDesc, err := storeJson(ctx, cs, ocispec.MediaTypeImageConfig, config, nil)
+	if err != nil {
+		return "", err
+	}
+
+	manifest := ocispec.Manifest{
+		Versioned: specs.Versioned{
+			SchemaVersion: 2,
+		},
+		Config: configDesc,
+		Layers: []ocispec.Descriptor{
+			compressedRootfsDesc,
+		},
+	}
+	manifestDesc, err := storeJson(ctx, cs, ocispec.MediaTypeImageManifest, manifest, map[string]string{
+		"containerd.io/gc.ref.content.config": configDesc.Digest.String(),
+		"containerd.io/gc.ref.content.l.0":    compressedDigest.String(),
+	})
+	if err != nil {
+		return "", err
+	}
+
+	id := image.ID(manifestDesc.Digest.String())
+	img := images.Image{
+		Name:      ref.String(),
+		Target:    manifestDesc,
+		CreatedAt: createdAt,
+	}
+	if img.Name == "" {
+		// TODO(vvoland): danglingImageName(manifestDesc.Digest)
+		img.Name = "dangling@" + manifestDesc.Digest.String()
+
+	}
+
+	err = i.saveAndUnpackImage(ctx, img, *platform)
+	if err != nil {
+		logger.WithError(err).Warn("failed to unpack image")
+	}
+
+	return id, err
+}
+
+// saveArchive saves the archive from bufRc to the content store, compressing it if necessary.
+// mediaType will be modified if the archive is decided to be compressed.
+func saveArchive(ctx context.Context, cs content.Store, mediaType *string, bufRc *bufio.Reader) (digest.Digest, digest.Digest, error) {
+	// If the input is already a compressed layer, just save it as is.
+	if *mediaType != ocispec.MediaTypeImageLayer {
+		compressedDigest, err := writeBlobAndReturnDigest(ctx, cs, *mediaType, bufRc)
+		if err != nil {
+			return "", "", errdefs.System(err)
+		}
+
+		uncompressedDigest, err := digestUncompressed(ctx, cs, compressedDigest)
```

Comment:
I don't think we can use t he TeeReader trick here, because decompressed stream is different Reader than the uncompressed one. 
I implemented it with another `io.Pipe` (meh), but maybe there's some clever way to work around this?